### PR TITLE
[kvutils] use new contract id scheme

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
In this PR we make kvutils work with the new random contract id scheme. 

In particular, we drop all relative contract of kvutils. 

This advances the state of #3830 

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [X] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto [hunk 1]
```diff
@@ -321,7 +324,7 @@ message DamlStateValue {
     DamlContractState contract_state = 2;
     DamlCommandDedupValue command_dedup = 3;
     DamlPartyAllocation party = 4;
-    DamlContractKeyState contract_key_state = 5;
+    string contract_key_state = 5;
```
Comment (dajmaki) [inline ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: Please keep it as a message so we can extend it if needed.

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto [hunk 2]
```diff
@@ -357,6 +360,9 @@ message DamlSubmissionDedupKey {
 
   // Uploading participant's id.
   string participant_id = 3;
+
+  // Seed use for interpretation.
+  bytes submissionSeed = 4;
```
Comment (dajmaki) [inline ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: Why is submission seed here? I don't see how this can work, since we must also be able to deduplicate commands that are sent by different participants (we support a deployment where we have clones of the participant for high availability).

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InputsAndEffects.scala [hunk 3]
```diff
@@ -106,57 +102,50 @@ private[kvutils] object InputsAndEffects {
...
                       (globalKeyToStateKey(
                         GlobalKey(
                           create.coinst.template,
-                          forceNoContractIds(keyWithMaintainers.key.value))) ->
-                        DamlContractKeyState.newBuilder
-                          .setContractId(encodeRelativeContractId(
-                            entryId,
-                            create.coid.asInstanceOf[RelativeContractId]))
-                          .build))
+                          // FIXME: We probably should not crash here.
```
Comment (dajmaki) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InputsAndEffects.scala]: preferably throw a kvutils `Err` here instead as the caller assumes only those are expected.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueConsumption.scala [hunk 4]
```diff
@@ -229,20 +229,15 @@ object KeyValueConsumption {
         workflowId = Some(txEntry.getWorkflowId)
           .filter(_.nonEmpty)
           .map(parseLedgerString("WorkflowId")),
+        submissionSeed = parseOptHash(txEntry.getSubmissionSeed)
       ),
-      transaction = makeCommittedTransaction(entryId, relTx),
+      transaction = relTx,
```
Comment (dajmaki) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueConsumption.scala]: rename `relTx`.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala [hunk 5]
```diff
@@ -35,7 +35,10 @@ object KeyValueSubmission {
     *
     * @deprecated Use [[KeyValueCommitting.submissionOutputs]] instead. This function will be removed in later version.
     */
-  def transactionOutputs(entryId: DamlLogEntryId, tx: SubmittedTransaction): List[DamlStateKey] = {
+  def transactionOutputs(
+      entryId: DamlLogEntryId,
```
Comment (dajmaki) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala]: Do we need `entryId` anymore? Preferrably we remove the whole `submissionOutputs` function and include the outputs in the submission message itself. This can and should probably be later work.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/api/KeyValueParticipantStateWriter.scala [hunk 6]
```diff
@@ -23,9 +23,14 @@ class KeyValueParticipantStateWriter(writer: LedgerWriter)(
...
       transactionMeta: TransactionMeta,
-      transaction: SubmittedTransaction): CompletionStage[SubmissionResult] = {
+      transaction: SubmittedTransaction,
+  ): CompletionStage[SubmissionResult] = {
     val submission =
-      KeyValueSubmission.transactionToSubmission(submitterInfo, transactionMeta, transaction)
+      KeyValueSubmission.transactionToSubmission(
+        submitterInfo,
+        transactionMeta,
+        transaction.assertNoRelCid(cid => s"Unexpected relative contract id: $cid"),
```
Comment (dajmaki) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/api/KeyValueParticipantStateWriter.scala]: I would rather catch the relative contract ids in the transaction encoding. Pretty expensive traversing it just for this.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala [hunk 7]
```diff
@@ -175,25 +180,23 @@ private[kvutils] case class ProcessTransactionSubmission(
...
       allUnique = relTx
         .fold((true, startingKeys)) {
           case (
               (allUnique, existingKeys),
-              (_nodeId, exe: NodeExercises[_, _, VersionedValue[ContractId]]))
+              (_nodeId, exe @ Node.NodeExercises(_, _, _, _, _, _, _, _, _, _, _, _, _, _)))
               if exe.key.isDefined && exe.consuming =>
             val stateKey = Conversions.globalKeyToStateKey(
-              GlobalKey(exe.templateId, Conversions.forceNoContractIds(exe.key.get.key.value)))
+              Node.GlobalKey(exe.templateId, Conversions.forceNoContractIds(exe.key.get.key.value)))
```
Comment (dajmaki) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala]: we can drop the `forceNoContractIds` as the GlobalKey constructor hashes it, which will throw if there's contract ids.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala [hunk 8]
```diff
@@ -60,23 +72,40 @@ object Hash {
   def assertFromBytes(a: Array[Byte]): Hash =
     data.assertRight(fromBytes(a))
 
-  def secureRandom(seed: Array[Byte]): () => Hash = {
-    val random = new SecureRandom(seed)
+  // A pseudo random generator for Hash based on hmac
+  // We mix the given seed with time to mitigate very bad seed.
+  def random(seed: Array[Byte]): () => Hash = {
+    if (seed.length != underlyingHashLength)
+      throw new IllegalArgumentException(s"expected a 32 bytes seed, get ${seed.length} bytes.")
```
Comment (miklos-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala]:
```suggestion
      throw new IllegalArgumentException(s"Expected a $underlyingHashLength bytes seed, got ${seed.length} bytes.")
```

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala [hunk 9]
```diff
@@ -136,15 +171,25 @@ object Hash {
...
+      val ss = set.toSeq.sorted[String]
+      iterateOver(ss.iterator, ss.size)(_ add _)
+    }
+
+    @throws[HashingError]
+    final def addCid(cid: Value.ContractId): this.type =
+      add(cid2String(cid))
+
     // In order to avoid hash collision, this should be used together
     // with an other data representing uniquely the type of `value`.
```
Comment (miklos-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala]:
```suggestion
    // with another data representing uniquely the type of `value`.
```

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto [hunk 10]
```diff
@@ -150,6 +150,9 @@ message DamlTransactionEntry {
   // the bounds for valid timestamps in relation to the ledger record time
   // (the time at which the transaction is committed).
   google.protobuf.Timestamp ledger_effective_time = 4;
+
+  // The nonce used to generate contract ids
+  bytes submissionSeed = 5;
```
Comment (dajmaki) [inline ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]:
```suggestion
  bytes submission_seed = 5;
```

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto (hunk 2 above)
Comment (dajmaki) [inline ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]:
```suggestion
  bytes submission_seed = 4;
```

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala [hunk 11]
```diff
@@ -192,6 +152,15 @@ private[state] object Conversions {
   def parseTimestamp(ts: com.google.protobuf.Timestamp): Time.Timestamp =
     Time.Timestamp.assertFromInstant(Instant.ofEpochSecond(ts.getSeconds, ts.getNanos.toLong))
 
+  def parseHash(a: com.google.protobuf.ByteString): crypto.Hash =
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala]: Please use meaningful argument names.

File: ledger/sandbox/src/test/lib/scala/com/digitalasset/platform/sandbox/ScenarioLoadingITBase.scala [hunk 12]
```diff
@@ -160,7 +160,7 @@ abstract class ScenarioLoadingITBase
             val contractIds = events.map(_.contractId).toSet
 
             // note how we skip #0, #2 and #5 because of the `pass`es in the scenario.
-            contractIds shouldBe Set("#4:0", "#7:2", "#3:0", "#7:1", "#1:0", "#7:0", "#6:0")
+            contractIds.size shouldBe 7 // Set("#4:0", "#7:2", "#3:0", "#7:1", "#1:0", "#7:0", "#6:0")
```
Comment (miklos-da) [inline ledger/sandbox/src/test/lib/scala/com/digitalasset/platform/sandbox/ScenarioLoadingITBase.scala]: Please remove commented out code and fix the above comment.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala [hunk 13]
```diff
@@ -217,20 +222,20 @@ final class Engine {
...
-        transactionSeed = tx.transactionSeed,
+        transactionSeed = transactionSeed,
         validating = true,
         checkSubmitterInMaintainers = checkSubmitterInMaintainers,
         submitters = submitters,
         commands = commands.map(_._2),
         time = ledgerEffectiveTime
       )
-      validationResult <- if (tx isReplayedBy rtx) {
+      validationResult <- if (tx isReplayedBy rtx)
```
Comment (miklos-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala]: Why change the formatting here?

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala (hunk 5 above)
Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala]: Does not seem with need it.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/api/KeyValueParticipantStateWriter.scala (hunk 6 above)
Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/api/KeyValueParticipantStateWriter.scala]: The objective is to completly get ride of the relative contract id. I would prefer kvutils type does not reference at all relative contract ids and drop support of old contract ids scheme in the sandbox as soon as possible (that is 2/3 releases). 
So we could simply merge `SubmittedTransaction` and `ComittedTransaction` (using only absolute contract id) and get rid of the `assertNoRelCid`

If it is very urgent we can find a work arround to drop completly relative contract ids and still have the sandbox support old scheme (that is `#txId:nodeId`)

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto (hunk 1 above)
Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: fixed

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto (hunk 2 above)
Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: It has nothing to do here. I dropped it. Good catch

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueConsumption.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueConsumption.scala]: renamed

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InputsAndEffects.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InputsAndEffects.scala]: done.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala (hunk 5 above)
Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueSubmission.scala]: dropped

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala (hunk 7 above)
Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala]: fixed

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto (hunk 10 above)
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto]: fixed

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala (hunk 8 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala]: fixed.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala (hunk 9 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala]: fixed

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala (hunk 11 above)
Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala]: fixed

File: ledger/sandbox/src/test/lib/scala/com/digitalasset/platform/sandbox/ScenarioLoadingITBase.scala (hunk 12 above)
Comment (remyhaemmerle-da) [reply ledger/sandbox/src/test/lib/scala/com/digitalasset/platform/sandbox/ScenarioLoadingITBase.scala]: fixed

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala (hunk 13 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala]: revert

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala [hunk 14]
```diff
@@ -50,6 +50,18 @@ object Hash {
 
   private val hexEncoding = BaseEncoding.base16().lowerCase()
 
+  private case class HashingError(msg: String) extends Error with NoStackTrace
```
Comment (gerolf-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala]: Do you really want to extend `java.lang.Error` and not `Exception`?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala [hunk 15]
```diff
@@ -345,14 +337,31 @@ object Node {
...
 
   /** Useful in various circumstances -- basically this is what a ledger implementation must use as
-    * a key. The 'hash' should be used when storing the key as value serialization is not guaranteed
-    * to be stable over time (e.g. a new encoding might be introduced).
+    * a key. The 'hash' is guaranteed to be stable over time.
     */
-  sealed abstract case class GlobalKey(templateId: Identifier, key: Value[Nothing], hash: Hash)
+  final class GlobalKey private (
+      val templateId: Identifier,
+      val key: Value[AbsoluteContractId],
```
Comment (gerolf-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala]: Why the change from `Value[Nothing]` to `Value[AbsoluteContractId]`?

File: ledger/sandbox/BUILD.bazel [hunk 16]
```diff
@@ -401,17 +401,34 @@ server_conformance_test(
     ],
 )
 
+server_conformance_test(
+    name = "conformance-test-contract-id-seeding",
```
Comment (gerolf-da) [inline ledger/sandbox/BUILD.bazel]: Is this an extra set of tests only for weak seeding?

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/SandboxServer.scala [hunk 17]
```diff
@@ -328,6 +320,8 @@ final class SandboxServer(
           ledgerType,
           authService.getClass.getSimpleName
         )
+        if (config.seeding.contains(SeedService.Seeding.Weak))
+          logger.warn("Contract id seeding uses a weak seed. DO NOT USE IN PRODUCTION.")
```
Comment (gerolf-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/SandboxServer.scala]: I don't think we need this warning to be honest. Maybe add the seeding-type to the startup info log, where we also print the supported LF versions.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandboxnext/Runner.scala [hunk 18]
```diff
@@ -119,6 +127,8 @@ class Runner {
           ledgerType,
           authService.getClass.getSimpleName,
         )
+        if (config.seeding.contains(SeedService.Seeding.Weak))
+          logger.warn("Contract id seeding uses a weak seed. DO NOT USE IN PRODUCTION.")
```
Comment (gerolf-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandboxnext/Runner.scala]: Same here, I don't think we need this warning.

File: ledger/sandbox/BUILD.bazel (hunk 16 above)
Comment (remyhaemmerle-da) [reply ledger/sandbox/BUILD.bazel]: By default Sandbox runs without seeding (on the contrary of kvutils), this test Sandbox works with seeding.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala (hunk 15 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala]: the middle plan is to get rid of `RelativeContractId`,  together with the parametrization of  `Value`.  Here the hashing of `key` done at the initialization of the `GlobalKey` object ensure there is no `ContractId` in `key`.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/api/KeyValueParticipantStateWriter.scala (hunk 6 above)
Comment (dajmaki) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/api/KeyValueParticipantStateWriter.scala]: sounds good to me. lets keep it and later clean up.
